### ISSUE TITLE
Stop using persisted connections on POST/PUT

### ISSUE THREAD

**Issue by pokeb** [pokeb/asi-http-request#94]
http://groups.google.com/group/asihttprequest/browse_thread/thread/7f75eabba015a657

**Comment by deltheil** [comment]
I've precisely experienced problems with persistent connections with POST/PUT requests, reproduced within an iPhone app context (the problem can be reproduced on both iPhone simulator and device, and on WiFi or 3G network).

The problem consists in an empty body received on the server-side, i.e. the server receives an HTTP header with `Content-Length: 0`. This problem has been observed with:
-   ASIHTTPRequest v1.8 (63061ac428a3ea639f4c)
-   nginx/0.7.67 configured with a Keep-Alive timeout equals to 35 seconds (`keepalive_timeout 35`)

I've found conditions to reproduce it **systematically**:
-   send a first asynchronous multipart POST request with default ASIHTTPRequest parameters (including persistent connections enabled and `persistentConnectionTimeoutSeconds` set to 60 seconds)
-   wait for more than 35 seconds (so that the server has closed the connection) and less than 60 seconds (so that the client will try to re-use the connection)
-   perform a second consecutive asynchronous multipart POST request: it **fails** because of empty data received on the server-side

When enabling all ASIHTTPRequest debugging traces, I systematically obtain such kind of traces for the 2nd request:

`Request #3 will use connection #2`

`Request attempted to use connection #2, but it has been closed - will retry with a new connection`

`Request #3 will use connection #3`

In other words the bug appears every time ASIHTTPRequest says it `will retry with a new connection`.

As a work around I've reduced the persistent connection timeout (`persistentConnectionTimeoutSeconds`) to a smaller value (< 35 seconds here) so that the client automatically closes the connection before it gets closed by the server.

**Comment by pokeb** [comment]
Hi Cédric

I've just done a quick test, and I'm not able to reproduce the above issue.

I tested with Apache - it returns a keep alive header to tell us how long to reuse the connection, so I had to disable the code that adjusts the timeout if it finds this header to ensure ASIHTTPRequest left the default persistent connection timeout on 60 seconds.

Here is my code:

```
NSURL *url = [NSURL URLWithString:@"http://myurl"];
ASIHTTPRequest *request = [ASIHTTPRequest requestWithURL:url];
[request setRequestMethod:@"PUT"];
[request appendPostData:[@"This is my body" dataUsingEncoding:NSUTF8StringEncoding]];
[request startSynchronous];
NSLog(@"%@",[request responseString]);

sleep(40);

request = [ASIHTTPRequest requestWithURL:url];
[request setRequestMethod:@"PUT"];
[request appendPostData:[@"This is my body" dataUsingEncoding:NSUTF8StringEncoding]];
[request startSynchronous];
NSLog(@"%@",[request responseString]);
```

The url it connects to simply returns the request body. Even though the second request was forced to retry on a new connection, the server still received the body.

If you are able to reproduce this on nginx with the same code, let me know, and I'll install nginx to do some testing.

retryUsingNewConnection really is supposed to be a fallback - the best approach would be one of:
- Set persistentConnectionTimeoutSeconds to the lower value, as you've done
- Configure the server to return a keep alive header, if this is possible
- Disable persistent connections for this request

I'd still like to change ASIHTTPRequest to turn off persistent connections for requests that have a body by default, since this seems to be a source of a lot of problems, and should help us better match browser behaviour.

Thanks

Ben

**Comment by deltheil** [comment]
Hi Ben,

I confirm I do **not** reproduce the problem with the exact same code and service.

That said, I observe the bug (with nginx + Keep-Alive set to 35 seconds) when working with a `multipart/form-data body`:

```
NSURL *url = [NSURL URLWithString:@"http://myurl"];
ASIFormDataRequest* request = [ASIFormDataRequest requestWithURL:url];
[request setRequestMethod:@"PUT"];
[request setData:[@"This is my body" dataUsingEncoding:NSUTF8StringEncoding] forKey:@"foo"];
[request startSynchronous];
NSLog(@"%@",[request responseString]);

sleep(40);

request = [ASIFormDataRequest requestWithURL:url];
[request setRequestMethod:@"PUT"];
[request setData:[@"This is my body" dataUsingEncoding:NSUTF8StringEncoding] forKey:@"foo"];
[request startSynchronous];
NSLog(@"%@",[request responseString]);
```

**Comment by Slyv** [comment]
I also have exactly the same problem with persistent connections as "deltheil".
It is possible that problem appears only on some servers.
Currently i'm using simple workaround:
[request setShouldAttemptPersistentConnection:NO];

How to reproduce:
Download project:
https://github.com/jaanus/PlainOAuth
After you login to Twitter, upload image to TwitPic. First time everything is OK.
But if you try to upload image second time, Asi-http-request will probably try to use persistent connection, and sending will fail.

debug messages:
First request:
Request #1 will use connection #1
Got no keep-alive header, will keep this connection open for 60.000000 seconds
Request #1 finished using connection #1

Second request:
Request #2 will use connection #1
Request attempted to use connection #1, but it has been closed - will retry with a new connection

AND HERE i have response from server:
"code":401,"message":"Could not authenticate you (no application key provided)"
It seems request's post-data key-values were cleared or something...

Request #2 will use connection #2
Got no keep-alive header, will keep this connection open for 60.000000 seconds
Request #2 finished using connection #2

**Comment by pokeb** [comment]
Ok, the above commit includes two changes:

1) Fix the problem that Cédric reported - requests were removing the file that stores a temporary request body (such as one generated by ASIFormDataRequest) before retrying

2) Persistent connections are now turned off by default for POST, PUT, and any other request that has a body at the time requestMethod is set. You can override this to use persistent connections for requests with a body by manually setting the requestMethod, then turning shouldAttemptPersistentConnection back on.

Thanks both for the info, it was very helpful in tracking this down. :)

Ben

**Comment by umairnow** [comment]
Hi Pokeb, I am facing the issue with sending json in post request, I am setting data but server is saying that post is empty. I am using this code, please help me out what can I do

```
ASIHTTPRequest *request = [ASIHTTPRequest requestWithURL:url];
//[request setRequestMethod:@"POST"];
[request addRequestHeader:@"Accept" value:@"application/json"];
[request addRequestHeader:@"content-type" value:@"application/x-www-form-urlencoded"];
[request setPostBody:[NSMutableData dataWithData:[jsonRequest  dataUsingEncoding:NSUTF8StringEncoding]]];
[request startSynchronous];
```

**Comment by ieswxia** [comment]
Your jsonRequest may not serialization.you can use some json framework to transform json object to binary data.datausingencode method cannot make it.

ÔÚ 2012-10-15£¬20:14£¬umairnow notifications@github.com Ð´µÀ£º

**Comment by umairnow** [comment]
yes I am using SBJSon actually when I use simple request and post data to server it works fine...it has only issue with ASIHTTPRequest

**Comment by ieswxia** [comment]
Thus,i think you should use asiformdatarequst instead.

ÔÚ 2012-10-15£¬21:27£¬umairnow notifications@github.com Ð´µÀ£º

**Comment by umairnow** [comment]
```
NSURL *url = [NSURL URLWithString:[urlStr stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
ASIFormDataRequest *request = [ASIFormDataRequest requestWithURL:url];
NSString *postJson = [NSString stringWithFormat:@"&json=%@", jsonRequest];

NSData *requestData = [postJson dataUsingEncoding:NSUTF8StringEncoding];
[request setPostBody:[NSMutableData dataWithData:requestData]];

[request startSynchronous];

NSDictionary *postResponse = [JasonParser parseJsonString:request.responseString];
```

I used this code but still server is unable to find post body

**Comment by ieswxia** [comment]
NSString *postJson = [NSString stringWithFormat:@"&json=%@", jsonRequest];
remove '&' ,'%@' and jsonRequest .then the server would receive post body.
I the type of jsonRequest,would you provide some more details?

ÔÚ 2012-10-15£¬22:06£¬umairnow notifications@github.com Ð´µÀ£º

**Comment by ieswxia** [comment]
I am sorry,i sended uncomplete email.
Actually,i do not understand the type of jsonRequst,can you explain it and show more details?

ÔÚ 2012-10-15£¬22:06£¬umairnow notifications@github.com Ð´µÀ£º

**Comment by umairnow** [comment]
Actually the server is responding fine by using this code

   NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60.0];

```
NSString *postJson = [NSString stringWithFormat:@"&json=%@", jsonRequest];

NSData *requestData = [postJson dataUsingEncoding:NSUTF8StringEncoding];

[request setHTTPMethod:@"POST"];
[request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
[request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"content-type"];
[request setValue:[NSString stringWithFormat:@"%d", [requestData length]] forHTTPHeaderField:@"Content-Length"];
[request setHTTPBody: requestData];

NSURLResponse *responseurl;
NSError *err;

NSData *data = [NSURLConnection sendSynchronousRequest:request returningResponse:&responseurl error:&err];
```
